src/common/grpc/src/flight.rs

```diff
@@ -0,0 +1,279 @@
+// Copyright 2022 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::pin::Pin;
+use std::sync::Arc;
+
+use api::result::ObjectResultBuilder;
+use api::v1::ObjectResult;
+use arrow_flight::utils::flight_data_to_arrow_batch;
+use arrow_flight::FlightData;
+use common_error::prelude::StatusCode;
+use common_recordbatch::{RecordBatch, RecordBatches};
+use datatypes::arrow::datatypes::Schema as ArrowSchema;
+use datatypes::arrow::ipc::{root_as_message, MessageHeader};
+use datatypes::schema::{Schema, SchemaRef};
+use futures::TryStreamExt;
+use prost::Message;
+use snafu::{OptionExt, ResultExt};
+use tonic::codegen::futures_core::Stream;
+use tonic::Response;
+
+use crate::error::{self, InvalidFlightDataSnafu, Result};
+
+type TonicResult<T> = std::result::Result<T, tonic::Status>;
+type TonicStream<T> = Pin<Box<dyn Stream<Item = TonicResult<T>> + Send + Sync + 'static>>;
+
+#[derive(Debug, Clone)]
+pub enum FlightMessage {
+    Schema(SchemaRef),
+    Recordbatch(RecordBatch),
+}
+
+#[derive(Default)]
+pub struct FlightDecoder {
+    schema: Option<SchemaRef>,
+}
+
+impl FlightDecoder {
+    pub fn try_decode(&mut self, flight_data: FlightData) -> Result<FlightMessage> {
```

Comment:
Why not try to decode from an iterator or stream of `FlightData`? This would eliminate the need for staging the schema.

Reply:
It would be more flexible because we could operate on a single FlightData element instead of the whole batch, helpful in writing tests; and it would be more easy to be extended in the future, other info could be stored in the decoder struct, acting like a "context".